### ISSUE TITLE
Compiling debug mode with -G results in ptxas resolution failure on 'cudaDeviceSynchronize'

### ISSUE THREAD

**Issue by nate-tri** [thrust/thrust#864]
Adjusting debug mode to turn on the -G (device debugging info) flag for nvcc results in a failure to compile for most targets when using an architecture >= 3.5

Tested with CUDA 8.0 and CUDA 7.5 on revision 7c7a5e6755ce18b988cc1284de36aa49513501e7

```
diff --git a/SConstruct b/SConstruct
index 5c1cdb2..c3f57a4 100644
--- a/SConstruct
+++ b/SConstruct
@@ -340,8 +340,7 @@ def nv_compiler_flags(mode, device_backend, arch, cdp):
   if mode == 'debug':
     # turn on debug mode
     # XXX make this work when we've debugged nvcc -G
-    #result.append('-G')
-    pass
+    result.append('-G')
   if device_backend != 'cuda':
     result.append("--x=c++")
   if cdp != False:
```

Result:
```
$ scons mode=debug -j 1 arch=sm_35 cuda_path=/usr/local/cuda-8.0/ 
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
nvcc -o performance/targets/cpp_host_cuda_device_debug_nvcc/adjacent_difference.o -c -gencode=arch=compute_35,\"code=sm_35,compute_35\" -G -Xcompiler -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -Xcompiler -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -Xcompiler -DTHRUST_DEBUG -Xcompiler -g -Xcompiler -Wall -Xcompiler -Werror -I . -I /usr/local/cuda-8.0//include -I performance -I testing performance/targets/cpp_host_cuda_device_debug_nvcc/adjacent_difference.cu
ptxas fatal   : Unresolved extern function 'cudaDeviceSynchronize'
scons: *** [performance/targets/cpp_host_cuda_device_debug_nvcc/adjacent_difference.o] Error 255
scons: building terminated because of errors.
```

**Comment by jiapei100** [comment]
There are 2 samples from 0_Simple requiring compute compability over 5.2.

In addition, I believe you have to make sure you libGL has been installed correctly and the symbolic link does link to something....

That's how I solve my problem under Ubuntu 16.04.1

Cheers
Pei

**Comment by nate-tri** [comment]
@jiapei100 - I was unable to make headway on the issue based on your suggestions.  My libGL appears to be correctly linked.

I'm also unclear as to how 0_Simple can require >5.2 when the revision referenced doesn't even support arch sm specifications >5.2.   If you meant arch sm == 5.2, then I can report that there is no change in behavior if arch=sm_52 is specified (and I have 5.2 hardware even).

Digging further, I was able to get a build going by doing the following:

```
diff --git a/SConstruct b/SConstruct
index 5c1cdb2..cc9f5eb 100644
--- a/SConstruct
+++ b/SConstruct
@@ -486,6 +485,7 @@ for (host,device) in itertools.product(host_backends, device
   if 'cudadevrt' in env['LIBS']:
     # nvcc is required to link against cudadevrt
     env.Replace(LINK = 'nvcc')
+    env.Replace(RPATHPREFIX = '-Xlinker=-rpath,')
 
     if os.name == 'nt':
       # the nv linker uses the same command line as the gnu linker
```

```
scons mode=debug -j 1 cdp=1 arch=sm_35 cuda_path=/usr/local/cuda-8.0/
```

This got me past the linker error, but running scons run_tests still fails with a segfault.

**Comment by brycelelbach** [comment]
Trying this out with CUDA 9.0

**Comment by brycelelbach** [comment]
Fixed in CUDA 9.0/Thrust 1.9.0. Right now, Thrust 1.9.0 is not pushed to Github; you'll have to get it from [CUDA 9.0](https://developer.nvidia.com/cuda-downloads).

However, the scons build system has languished a bit in the past year or so before I came to NVIDIA. I had to hack it up to get it working.

Internally, we use a (few) different build system(s) and test driver(s), and the scons system hasn't been kept up to date. I anticipate some sort of future unification to a single system for my own sanity.
